### PR TITLE
switch from jwt to PyJWT

### PR DESCRIPTION
**Pull request by groovecoder**

### CODE REVIEW

**Comment by groovecoder** [comment]
Note: this is causing https://sentry.prod.mozaws.net/operations/fx-private-relay-dev/issues/9682934/activity/?query=is%3Aunresolved until I figure out how exactly to parse and decode the FXA verifying public key.

**Comment by groovecoder** [comment]
Okay, I finally got this into a state where the branch worked on dev.fxprivaterelay.nonprod.cloudops.mozgcp.net. Ready for code review too.

File: privaterelay/views.py [hunk 1]
```diff
@@ -151,25 +150,40 @@ def fxa_rp_events(request):
...
-        return jwt_instance.decode(jwt, verifying_key)
+    authentic_jwt = _verify_jwt_with_fxa_key(req_jwt, private_relay_config)
 
+    if not authentic_jwt:
+        # FXA key may be old? re-fetch FXA keys and try again
+        private_relay_config.ready()
+        authentic_jwt = _verify_jwt_with_fxa_key(req_jwt, private_relay_config)
+        if not authentic_jwt:
+            raise Exception("Could not authenticate JWT with FXA key.")
+        return authentic_jwt
```
Comment (say-yawn) [inline privaterelay/views.py:166]:
```suggestion
```
return handled by outer `return` on line 168

**Review by say-yawn — CHANGES_REQUESTED** [review]
Minor changes

File: privaterelay/views.py [hunk 2]
```diff
@@ -151,25 +150,40 @@ def fxa_rp_events(request):
...
-        verifying_key = jwk_from_dict(verifying_key_json)
-        return jwt_instance.decode(jwt, verifying_key)
+    authentic_jwt = _verify_jwt_with_fxa_key(req_jwt, private_relay_config)
 
+    if not authentic_jwt:
+        # FXA key may be old? re-fetch FXA keys and try again
+        private_relay_config.ready()
+        authentic_jwt = _verify_jwt_with_fxa_key(req_jwt, private_relay_config)
+        if not authentic_jwt:
+            raise Exception("Could not authenticate JWT with FXA key.")
```
[the remark this answers is not in the document]
Comment (groovecoder) [reply privaterelay/views.py:165]: I think I started that way, but I had to refactor the retry and raise logic "up" into a containing function, and created the new specific `_verify_jwt_with_fxa_key` function so I could call it twice without duplicating all the code inside it. E.g., the code started like this:
```python
def _authenticate_fxa_jwt(req_jwt):
    private_relay_config = apps.get_app_config('privaterelay')
    social_app = SocialApp.objects.get(provider='fxa')
    authentic_jwt = None
    for verifying_key in private_relay_config.fxa_verifying_keys:
        if verifying_key['alg'] == 'RS256':
            verifying_key = jwt.algorithms.RSAAlgorithm.from_jwk(
                json.dumps(verifying_key)
            )
            authentic_jwt = jwt.decode(
                req_jwt,
                verifying_key,
                audience=social_app.client_id,
                algorithms=['RS256']
            )

    if not authentic_jwt:
        # FXA key may be old? re-fetch FXA keys and try again
        private_relay_config.ready()
        for verifying_key in private_relay_config.fxa_verifying_keys:
            if verifying_key['alg'] == 'RS256':
                verifying_key = jwt.algorithms.RSAAlgorithm.from_jwk(
                    json.dumps(verifying_key)
                )
                authentic_jwt = jwt.decode(
                    req_jwt,
                    verifying_key,
                    audience=social_app.client_id,
                    algorithms=['RS256']
                )

        if not authentic_jwt:
            raise Exception("Could not authenticate JWT with FXA key.")

    return authentic_jwt
```

And then I refactored into a `_verify_jwt_with_fxa_key` to reduce to code duplication:

```python
def _authenticate_fxa_jwt(req_jwt):
    private_relay_config = apps.get_app_config('privaterelay')
    authentic_jwt = _verify_jwt_with_fxa_key(req_jwt, private_relay_config)

    if not authentic_jwt:
        # FXA key may be old? re-fetch FXA keys and try again
        private_relay_config.ready()
        authentic_jwt = _verify_jwt_with_fxa_key(req_jwt, private_relay_config)
        if not authentic_jwt:
            raise Exception("Could not authenticate JWT with FXA key.")

    return authentic_jwt


def _verify_jwt_with_fxa_key(req_jwt, private_relay_config):
    social_app = SocialApp.objects.get(provider='fxa')
    for verifying_key in private_relay_config.fxa_verifying_keys:
        if verifying_key['alg'] == 'RS256':
            verifying_key = jwt.algorithms.RSAAlgorithm.from_jwk(
                json.dumps(verifying_key)
            )
            return jwt.decode(
                req_jwt,
                verifying_key,
                audience=social_app.client_id,
                algorithms=['RS256']
            )
```

Maybe the function names could be clearer?

File: privaterelay/views.py [hunk 3]
```diff
@@ -151,25 +150,40 @@ def fxa_rp_events(request):
...
+            verifying_key = jwt.algorithms.RSAAlgorithm.from_jwk(
+                json.dumps(verifying_key)
+            )
+            return jwt.decode(
+                req_jwt,
+                verifying_key,
+                audience=social_app.client_id,
+                algorithms=['RS256']
+            )
+
```
[the remark this answers is not in the document]
Comment (groovecoder) [reply privaterelay/views.py:183]: If this ever returns `None`, then we'll see `Could not authenticate JWT with FXA key.` exception raised from `_authenticate_fxa_jwt`. Which I think has the same effect.

File: privaterelay/views.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (say-yawn) [reply privaterelay/views.py:165]: No I think this is fine. I was thinking if there was a cleaner way to retry and was bouncing the idea here 😓 One day we'll figure out how to design retries better, but today isn't the day.

**Review by say-yawn — APPROVED** [review]
LGTM

**Pull request merged**